### PR TITLE
CRP Vehcile Controller integration 

### PR DESCRIPTION
**Pull request by mesmatyi**
PR for the **ctrl_vehicle_control** package

### CODE REVIEW

File: crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            m_targetCurvature = std::max(std::min(m_targetCurvature, params.maxCurvature), -params.maxCurvature);
+            if(params.debugKPIs){printf("kappa after limitation %f\n", m_targetCurvature);}
+            // the target curvature is compensated with a velocity dependent scaling factor
+            m_targetCurvature = m_targetCurvature*(params.ffGainOffsetGround+input.vxEgo*params.ffGainSlope);
+
+            if(params.debugKPIs){printf("kappa after velocity compensation %f\n", m_targetCurvature);}
+            // finally, lateral feedforward acceleration is calculated - this is done to be in harmony 
+            // with the feedback path later (out of scope of this method)
+            m_targetAccelerationFF = m_targetCurvature*std::pow(input.vxEgo,2);
+            if(params.debugKPIs){printf("ayTarFF %f\n", m_targetAccelerationFF);}
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp:107]: These lines could use some new lines and tabs/spaces.

File: crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            m_posErrPrev = m_posErr;
+        }
+
+        void compensatoryModel::cutRelevantLocalSnippet()
+        {
+            // this method cuts a relevant length of the total trajectory, based on the complete 
+            // path transformed to the ego coordinate frame
+            // step 1: searching for nearest point in trajectory
+            unsigned long int startIdx = -1;
+            unsigned long int stopIdx = -1;
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp:160]: Consider changing this, it looks a bit weird for me.
```suggestion
            unsigned long int stopIdx = std::numeric_limits<unsigned long int>::max();;
```

File: crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+#include "ctrl_vehicle_control/controller_inputs.hpp"
+#include "ctrl_vehicle_control/controller_outputs.hpp"
+#include "lib/utils.hpp"
+#include "compensatory_model/compensatory_model.hpp"
+
+using namespace std::chrono_literals;
+using std::placeholders::_1;
+
+class ctrl_vehicle_control : public rclcpp::Node
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp:28]:
```suggestion
class CtrlVehicleControl : public rclcpp::Node
```
Also definitions should be in a header file (.hpp)

File: crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            input.path_x.push_back(input_msg.points.at(i).pose.position.x);
+            input.path_y.push_back(input_msg.points.at(i).pose.position.y);
+            quaternion[0] = input_msg.points.at(i).pose.orientation.x; 
+            quaternion[1] = input_msg.points.at(i).pose.orientation.y; 
+            quaternion[2] = input_msg.points.at(i).pose.orientation.z; 
+            quaternion[3] = input_msg.points.at(i).pose.orientation.w; 
+            input.path_theta.push_back(m_geometricOperator.transformQuatToEuler(quaternion));
+            if (i<10){
+                //printf("input trajectory %d: %f %f %f\n", i, input_msg.points.at(i).pose.position.x, input_msg.points.at(i).pose.position.y, m_geometricOperator.transformQuatToEuler(quaternion));
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp:83]: This should be un-commented or removed

File: crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+private:
+    // VARIABLES
+    crp::apl::controlInput input;
+    crp::apl::controlOutput output;
+    crp::apl::controlParams params;
+    crp::apl::geometricOperators m_geometricOperator;
+    crp::apl::compensatoryModel m_compensatoryModel;
+
+    void traj_callback(const autoware_auto_planning_msgs::msg::Trajectory input_msg)
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp:65]:
```suggestion
    void trajCallback(const autoware_auto_planning_msgs::msg::Trajectory input_msg)
```

File: crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            if (i<10){
+                //printf("input trajectory %d: %f %f %f\n", i, input_msg.points.at(i).pose.position.x, input_msg.points.at(i).pose.position.y, m_geometricOperator.transformQuatToEuler(quaternion));
+            }
+        }
+
+        input.target_speed = input_msg.points.at(0).longitudinal_velocity_mps;
+   
+    }
+
+    void ego_vehicle_callback(const crp_msgs::msg::Ego input_msg)
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp:91]:
```suggestion
    void egoVehicleCallback(const crp_msgs::msg::Ego input_msg)
```

File: crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+        cmd_pub->publish(ctrl_cmd);
+    }
+    rclcpp::TimerBase::SharedPtr timer_;
+    rclcpp::Publisher<autoware_auto_control_msgs::msg::AckermannControlCommand>::SharedPtr cmd_pub;
+    
+    rclcpp::Subscription<autoware_auto_planning_msgs::msg::Trajectory>::SharedPtr traj_sub;
+    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub;
+    rclcpp::Subscription<crp_msgs::msg::Ego>::SharedPtr ego_vehicle_sub;
+    rclcpp::Subscription<autoware_auto_control_msgs::msg::AckermannControlCommand>::SharedPtr cmd_dummy_sub;
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/src/ctrl_vehicle_control.cpp:140]: Missing m_ prefix and camelCase with member variables
```suggestion
    rclcpp::Subscription<autoware_auto_control_msgs::msg::AckermannControlCommand>::SharedPtr m_sub_cmdDummy;
```

File: crp_APL/ctrl_vehicle_control/include/lib/utils.hpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#include <cmath>
+#include <stdio.h>
+
+#ifndef UTILS_HPP_
+#define UTILS_HPP_
```
[the remark this answers is not in the document]
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/include/lib/utils.hpp:2]: Done

File: crp_APL/ctrl_vehicle_control/package.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
+  <depend>rclcpp</depend>
+  <depend>geometry_msgs</depend>
+  <depend>std_msgs</depend>
+  <depend>autoware_auto_control_msgs</depend>
+  <depend>autoware_auto_planning_msgs</depend>
+  <depend>autoware_auto_vehicle_msgs</depend>
+  <depend>nav_msgs</depend>
+  <depend>pacmod3_msgs</depend>
```
[the remark this answers is not in the document]
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/package.xml:23]: Added

File: crp_APL/ctrl_vehicle_control/launch/mini_sim-launch.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            {"/ctrl/ffGainOffsetGround": 1.0},
+            {"/ctrl/ffGainSlope": 0.1},
+            {"/ctrl/ffLookAheadTime": 1.5},
+            {"/ctrl/steeringAngleLPFilter": 0.7},
+            {"/ctrl/fbLookAheadTime": 0.5},
+            {"/ctrl/fbPGain": 1.0},
+            {"/ctrl/fbDGain": 0.1},
+            {"/ctrl/fbIGain": 0.01},
+            {"/ctrl/fbMinLookAheadDistance": 0.0},
+            {"/ctrl/fbIntegralLimit": 3.0}
```
[the remark this answers is not in the document]
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/launch/mini_sim-launch.py]: For now, just keep it simple, but later for sure....

File: crp_APL/ctrl_vehicle_control/launch/run_all.launch.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+            {"/ctrl/ffGainOffsetGround": 1.0},
+            {"/ctrl/ffGainSlope": 0.1},
+            {"/ctrl/ffLookAheadTime": 1.5},
+            {"/ctrl/steeringAngleLPFilter": 0.7},
+            {"/ctrl/fbLookAheadTime": 0.5},
+            {"/ctrl/fbPGain": 0.01},
+            {"/ctrl/fbDGain": 0.0},
+            {"/ctrl/fbIGain": 0.0},
+            {"/ctrl/fbMinLookAheadDistance": 0.0},
+            {"/ctrl/fbIntegralLimit": 3.0}
```
[the remark this answers is not in the document]
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/launch/run_all.launch.py:20]: Kis

File: crp_APL/ctrl_vehicle_control/include/ctrl_vehicle_control/controller_inputs.hpp [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+            double steeringAngleLPFilter{0.7f};
+            double fbLookAheadTime{0.0f};
+            double fbPGain{0.5f};
+            double fbDGain{0.1f};
+            double fbIGain{0.01f};
+            double fbThetaGain{0.0f};
+            double fbMinLookAheadDistance{0.0f};
+            double fbIntegralLimit{3.0f};
+            double invSteerTimeConstant{0.0f}; // todo
+            double invSteerDamping{0.0f}; // todo
```
[the remark this answers is not in the document]
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/include/ctrl_vehicle_control/controller_inputs.hpp:40]: Currently the function that would use these params is not used...

File: crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp (hunk 2 above)
Comment (mesmatyi) [reply crp_APL/ctrl_vehicle_control/src/compensatory_model/compensatory_model.cpp:157]: max -ing out this is a bit risky, if the stop idx remains the same the indexing would run out

**Comment by mesmatyi** [comment]
Done with the required changes @AnonymDavid take a look again

File: crp_APL/ctrl_vehicle_control/README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+| /ctrl/ffGainOffsetGround    | double | 1.0                                             |
+| /ctrl/ffGainSlope | double | 0.1 |
+| /ctrl/ffLookAheadTime | double | 1.5
+| /ctrl/steeringAngleLPFilter | double | 0.7
+| /ctrl/fbLookAheadTime | double | 0.5
+| /ctrl/fbPGain | double | 0.01
+| /ctrl/fbDGain | double | 0.0
+| /ctrl/fbIGain | double | 0.0
+| /ctrl/fbMinLookAheadDistance | double | 0.0
+| /ctrl/fbIntegralLimit | 3.0
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/README.md:27]: Type is missing

File: crp_APL/ctrl_vehicle_control/package.xml [hunk 14]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>ctrl_vehicle_control</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
```
Comment (AnonymDavid) [inline crp_APL/ctrl_vehicle_control/package.xml:6]: Description missing

**Review by AnonymDavid — COMMENTED** [review]
Documentation is missing from SystemSpecification.tex/pdf
I would put the autoware_msgs and the tier4 package into a folder (e.g. external).
Other these that and some questions it looks good.

**Comment by mesmatyi** [comment]
Moved the subrepos to **external**and added documentation in tex,pdf and README
@AnonymDavid @gfigneczi1

**Pull request merged**